src/mrpro/operators/WaveletOp.py

```diff
@@ -0,0 +1,359 @@
+
+        # move stacked coefficients to first wavelet dimension
+        return (coefficients_stack,)
+
+    def adjoint(self, coefficients_stack: torch.Tensor) -> tuple[torch.Tensor]:
+        """Transform wavelet coefficients to (image) data.
+
+        Parameters
+        ----------
+        coefficients_stack
+            Wavelet coefficients stacked along one dimension
+
+        Returns
+        -------
+            (Image) data
+
+        Raises
+        ------
+        ValueError
+            If the domain_shape is not defined.
+        ValueError
+            If the dimensions along which wavelets are to be calculated are not unique.
+        """
+        if self.domain_shape is None:
+            raise ValueError('Adjoint requires to define the domain_shape in init()')
+
+        # normalize axes to allow negative indexing in input
+        dim = tuple(d % (coefficients_stack.ndim + len(self.dim) - 1) for d in self.dim)
+        if len(dim) != len(set(dim)):
+            raise ValueError(f'Axis must be unique. Normalized axis are {dim}')
+
+        coefficients_stack = torch.moveaxis(coefficients_stack, min(dim), -1)
+
+        # the ptwt functions work only for real data, thus we handle complex inputs as an additional channel
+        coefficients_stack_real = (
+            torch.view_as_real(coefficients_stack).moveaxis(-1, 0)
+            if coefficients_stack.is_complex()
+            else coefficients_stack
+        )
+
+        coefficients_list = self._stacked_tensor_to_coeff(coefficients_stack_real)
+
+        if len(self.dim) == 1:
+            coeffs1 = self._undo_format_coeffs_1d(coefficients_list)
+            data = waverec(coeffs1, self.wavelet_name, axis=-1)
+        elif len(self.dim) == 2:
+            coeffs2 = self._undo_format_coeffs_2d(coefficients_list)
+            data = waverec2(coeffs2, self.wavelet_name, axes=(-2, -1))
+        elif len(self.dim) == 3:
+            coeffs3 = self._undo_format_coeffs_3d(coefficients_list)
+            data = waverec3(coeffs3, self.wavelet_name, axes=(-3, -2, -1))
+
+        # undo swapping of axes
+        if coefficients_stack.is_complex():
+            data = torch.moveaxis(
+                data, list(range(-len(self.dim), 0)), [d + 1 for d in dim]
+            )  # +1 because first dim is real/imag
+            # if we deal with complex coefficients, we also return complex data
+            data = torch.view_as_complex(data.moveaxis(0, -1).contiguous())
+        else:
+            data = torch.moveaxis(data, list(range(-len(self.dim), 0)), dim)
+
+        return (data,)
+
+    def _format_coeffs_1d(self, coefficients: list[torch.Tensor]) -> list[torch.Tensor]:
+        """Format 1D wavelet coefficients to MRpro format.
+
+        At the moment, this function just returns the input coefficients as is:
+        [a, d_n, ..., d_1]
+        """
+        return coefficients
+
+    def _undo_format_coeffs_1d(self, coefficients: list[torch.Tensor]) -> list[torch.Tensor]:
```

Comment:
can these all be static if they don't access anything from self? or outside the class (in the same file)